### PR TITLE
adding word break DP approach

### PR DESCRIPTION
**Pull request by axayjha**
#### Description of Change
Adding dynamic programming approach for the famous Word Break problem which has been frequently asked in FAANG interviews.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1278"></a>

### CODE REVIEW

File: dynamic_programming/word_break.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        vector<int> dp(s.length(), INT_MAX);
+        return check(s, strSet, 0, &dp);
+    }
+};
+
+int main() {
+    const string s = "applepenapple";
+    const vector<string> wordDict = {"apple", "pen"};
+    cout << Solution().wordBreak(s, wordDict) << endl;
+}
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:82]:
```suggestion
}

```

File: dynamic_programming/word_break.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        for (const auto &s : wordDict) {
+            strSet.insert(s);
+        }
+
+        vector<int> dp(s.length(), INT_MAX);
+        return check(s, strSet, 0, &dp);
+    }
+};
+
+int main() {
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:78]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Documentation is not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
Please read them **carefully** and **follow them**.

**Comment by axayjha** [comment]
Have made the requested changes. Added documentation and enhancements.

File: dynamic_programming/word_break.cpp (hunk 1 above)
Comment (axayjha) [reply dynamic_programming/word_break.cpp]: Done

File: dynamic_programming/word_break.cpp (hunk 2 above)
Comment (axayjha) [reply dynamic_programming/word_break.cpp:141]: Done

File: dynamic_programming/word_break.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+#include <unordered_set>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::string;
+using std::unordered_set;
+using std::vector;
+
+/**
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:39]:
```suggestion
/**
 * @namespace dynamic_programming
 * @brief Dynamic programming algorithms
 */
namespace dynamic_programming {
/**
 * @namespace word_break
 * @brief Functions for Word Break algorithm
 */
namespace word_break {
/**
```

File: dynamic_programming/word_break.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        // tell if the string s can be segmented (correctly) at position i.
+        // initializing it with INT_MAX (which will denote no solution)
+        vector<int> dp(s.length(), INT_MAX);
+
+        // calling check method with position = 0, to check from left
+        // from where can be start segmenting the complete string in correct
+        // manner
+        return check(s, strSet, 0, &dp);
+    }
+};
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:135]:
```suggestion
};
}  // namespace word_break
}  // namespace dynamic_programming

```

File: dynamic_programming/word_break.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+ */
+int main() {
+    // the complete string
+    const string s = "applepenapple";
+    // the dictionary to be used
+    const vector<string> wordDict = {"apple", "pen"};
+
+    // should return true, as applepenapple can be segmented as apple + pen +
+    // apple
+    cout << Solution().wordBreak(s, wordDict) << endl;
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:149]:
```suggestion
    cout << dynamic_programming::word_break::Solution().wordBreak(s, wordDict) << endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/axayjha/C-Plus-Plus/actions

**Comment by axayjha** [comment]
Added Github Actions to my fork

**Comment by axayjha** [comment]
Have made the requested changes

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by axayjha — COMMENTED** [review]
Code is now conforming to guidelines

File: dynamic_programming/word_break.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
+/**
+ * @file
+ * @brief [Word Break Problem] (https://leetcode.com/problems/word-break/)
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:3]:
```suggestion
 * @brief [Word Break Problem](https://leetcode.com/problems/word-break/)
```

File: dynamic_programming/word_break.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic programming algorithms
+ */
+namespace dynamic_programming {
+
+/**
+ * @namespace word_break
+ * @brief Functions for Word Break algorithm
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:45]:
```suggestion
 * @brief Functions for [Word Break](https://leetcode.com/problems/word-break/) problem
```

File: dynamic_programming/word_break.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+namespace word_break {
+
+/**
+ * @brief Function that checks if the string passed in param is present in
+ * the the unordered_set passed
+ *
+ * @param str the string to be searched
+ * @param strSet unordered set of string, that is to be looked into
+ * @returns `true` if str is present in strSet
+ * @returns `false` otherwise
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:56]:
```suggestion
 * @returns `false` if str is not present in strSet
```

File: dynamic_programming/word_break.cpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+ * Output: false
+ *
+ * @author [Akshay Anand] (https://github.com/axayjha)
+ */
+
+#include <climits>
+#include <iostream>
+#include <string>
+#include <unordered_set>
+#include <vector>
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:35]:
```suggestion
#include <vector>
#include <cassert>
```

File: dynamic_programming/word_break.cpp [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+    // calling check method with position = 0, to check from left
+    // from where can be start segmenting the complete string in correct
+    // manner
+    return check(s, strSet, 0, &dp);
+}
+
+}  // namespace word_break
+}  // namespace dynamic_programming
+
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:149]:
```suggestion

/**
 * @brief Test implementations
 * @returns void
 */
static void test() {
    // the complete string
    const std::string s = "applepenapple";
    // the dictionary to be used
    const std::vector<std::string> wordDict = {"apple", "pen"};

    assert(dynamic_programming::word_break::wordBreak(s, wordDict));

    // should return true, as applepenapple can be segmented as apple + pen +
    // apple
    std::cout << dynamic_programming::word_break::wordBreak(s, wordDict)
              << std::endl;
    std::cout << "Test implementation passed!\n";
 }
```

File: dynamic_programming/word_break.cpp [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+int main() {
+    // the complete string
+    const std::string s = "applepenapple";
+    // the dictionary to be used
+    const std::vector<std::string> wordDict = {"apple", "pen"};
+
+    // should return true, as applepenapple can be segmented as apple + pen +
+    // apple
+    std::cout << dynamic_programming::word_break::wordBreak(s, wordDict)
+              << std::endl;
```
Comment (Panquesito7) [inline dynamic_programming/word_break.cpp:163]:
```suggestion
    test(); // run the tests
    return 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work, code looks pretty refined now. 😄 👍
Let us add some tests using `assert` to check if the code runs fine.

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've improved the code so well.
Code and documentation looks good and very refined; LGTM. 😄 👍

**Comment by axayjha** [comment]
Thanks man, got to learn a lot in the process. Cheers

**Review by kvedala — APPROVED** [review]
👍 Well written and documented code

File: dynamic_programming/word_break.cpp [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    std::cout << "Test implementation passed!\n";
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    // the complete string
```
Comment (kvedala) [inline dynamic_programming/word_break.cpp:175]:
```suggestion

   test(); // call the test function :)

    // the complete string
```

**Comment by axayjha** [comment]
@kvedala @Panquesito7 
Hey chief, would you be kind enough to please label this PR as 'hacktoberfest-accepted'?

**Comment by Panquesito7** [comment]
> label this PR as 'hacktoberfest-accepted'?

This repository has the `hacktoberfest` topic, so, once your PR is
merged, it counts toward completing the Hacktoberfest challenge. 🙂

**Comment by axayjha** [comment]
Yes, I noticed after commenting. 😜  Thanks. 😄

**Pull request merged**